File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Generate Parentheses/Generate Parentheses.cpp
class Solution {
public:
    // this is my own thought process
    void helper(vector<string> &ans, string str, int opens, int closes) {
        if (opens == 0 || closes == 0) {
            for (int i=0; i<closes; ++i) str.push_back(')');
            ans.push_back(str);
            return;
        }
        helper(ans, str+"(", opens-1, closes);
        if (opens < closes)
            helper(ans, str+")", opens, closes - 1);
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        helper(ans, "(", n-1, n);
        return ans;
    }
};

File: Step 07 Recursion/Step 7.2 Subsequence Pattern/Generate Parentheses/Generate Parentheses.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        helper(result, "", n, n);
        return result;
    }

    private void helper(List<String> result, String str, int opens, int closes) {
        if (opens == 0 && closes == 0) {
            result.add(str);
            return;
        }
        if (opens > 0) {
            helper(result, str + "(", opens - 1, closes);
        }
        if (closes > opens) {
            helper(result, str + ")", opens, closes - 1);
        }
    }
}